raiden/network/transport/matrix/client.py

```diff
@@ -432,38 +447,41 @@ def get_user_presence(self, user_id: str) -> str:
     def call(callback, *args, **kwargs):
         return callback(*args, **kwargs)
 
-    def _sync(self, timeout_ms=30000):
+    def _sync(self, timeout_ms=SYNC_TIMEOUT_MS):
         """ Reimplements MatrixClient._sync, add 'account_data' support to /sync """
         log.debug(
             "Sync called", node=node_address_from_userid(self.user_id), current_user=self.user_id
         )
 
         response = self.api.sync(self.sync_token, timeout_ms)
-        prev_sync_token = self.sync_token
-        self.sync_token = response["next_batch"]
-
-        if self._handle_thread is not None:
-            # if previous _handle_thread is still running, wait for it and re-raise if needed
-            self._handle_thread.get()
 
-        is_first_sync = prev_sync_token is None
-        self._handle_thread = gevent.Greenlet(self._handle_response, response, is_first_sync)
-        self._handle_thread.name = (
-            f"GMatrixClient._sync user_id:{self.user_id} sync_token:{prev_sync_token}"
-        )
-        self._handle_thread.link_exception(lambda g: self.sync_thread.kill(g.exception))
-        log.debug(
-            "Starting handle greenlet",
-            node=node_address_from_userid(self.user_id),
-            first_sync=is_first_sync,
-            sync_token=prev_sync_token,
-            current_user=self.user_id,
-        )
-        self._handle_thread.start()
+        self.response_queue.put(response)
+        self.sync_token = response["next_batch"]
 
         if self._post_hook_func is not None:
             self._post_hook_func(self.sync_token)
 
+    def _handle_process_worker(self, response_queue: NotifyingQueue, event_stop: Event) -> None:
+        data_or_stop = event_first_of(response_queue, event_stop)
+
+        while True:
+            data_or_stop.wait()
+
+            # Clear the event's internal state for the next iteration.
+            #
+            # This *must* be done before any context switch, otherwise cuncurrent
+            # actions can be missed.
+            data_or_stop.clear()
+
+            if event_stop.is_set():
+                return
+
+            # The queue is not empty at this point, so this won't raise Empty.
+            response: Dict = response_queue.get(block=False)
```

Comment:
This needs to be fixed, if the thread is `stop`ed after this call, but before the complete response is processed, messages can be lost. The correct approach here is to peek the queue and only do the `get` once `_handle_response` returns.